vclib/src/commonMain/kotlin/at/asitplus/wallet/lib/data/dif/PresentationPreparationHelper.kt

```diff
@@ -0,0 +1,99 @@
+    val fallbackFormatHolder: FormatHolder? = null,
+) {
+    constructor(
+        presentationDefinition: PresentationDefinition,
+        fallbackFormatHolder: FormatHolder? = null,
+    ) : this(
+        presentationDefinitionId = presentationDefinition.id,
+        submissionRequirements = presentationDefinition.submissionRequirements,
+        privateInputDescriptorMatches = presentationDefinition.inputDescriptors.associateWith {
+            // need to run a refresh against the credential store after initialization
+            listOf<HolderAgent.CandidateInputMatchContainer>()
+        },
+        fallbackFormatHolder = presentationDefinition.formats ?: fallbackFormatHolder,
+    )
+
+    val inputDescriptorMatches by ::privateInputDescriptorMatches
+    val inputDescriptorGroups: Map<String, String>
+        get() = inputDescriptorMatches.keys.associate {
+            it.id to it.group!! // group is checked at code point:7fe1e939-913c-47c9-9309-1e0d3ea39a9c
+        }
+
+    init {
+        if(submissionRequirements != null) {
+            inputDescriptorMatches.entries.forEach {
+                if (it.key.group == null) {
+                    // code point:7fe1e939-913c-47c9-9309-1e0d3ea39a9c
+                    throw MissingInputDescriptorGroupException(it.key)
+                }
+            }
+        }
+    }
+
+    suspend fun refreshInputDescriptorMatches(
+        holder: Holder,
+        pathAuthorizationValidator: PathAuthorizationValidator?,
+    ) {
+        privateInputDescriptorMatches = holder.matchInputDescriptorsAgainstCredentialStore(
+            inputDescriptors = inputDescriptorMatches.keys,
+            fallbackFormatHolder = fallbackFormatHolder,
+            pathAuthorizationValidator = pathAuthorizationValidator,
+        ).getOrThrow().mapValues {
+            it.value.toList()
+        }
+    }
+
+    fun isSubmissionRequirementsSatisfied(
+        submittedInputDescriptorIds: Set<String>,
+    ): Boolean {
+        return if (!isSubmissionRequirementListSatisfied(submittedInputDescriptorIds)) {
+            false
+        } else {
+            // do not allow submissions for unnecessary input descriptors
+            findUnnecessaryInputDescriptorSubmissions(submittedInputDescriptorIds).isEmpty()
+        }
+    }
+
+    fun findUnnecessaryInputDescriptorSubmissions(submittedInputDescriptorIds: Set<String>): Set<String> {
+        return submittedInputDescriptorIds.filter {
+            isSubmissionRequirementListSatisfied(submittedInputDescriptorIds - it)
+        }.toSet()
+    }
+
+    private fun isSubmissionRequirementListSatisfied(
+        submittedInputDescriptorIds: Set<String>,
+    ): Boolean = submissionRequirements?.all {
+        it.evaluate(
+            inputDescriptorGroups = inputDescriptorMatches.keys.associate {
+                it.id to it.group!! // group is checked at code point:7fe1e939-913c-47c9-9309-1e0d3ea39a9c
```

Comment:
What does this comment mean?

Reply:
As there is a double exclamation that the group is not null, I felt the need to explain the reasoning behind it

There is another part in the class, where the group is checked in case of existing submission requirements. This validation can be found when searching for the string "7fe1e939-913c-47c9-9309-1e0d3ea39a9c". 

changed for better readability in commit c0944dcdee8532eb1c6e0d081798eb7c0370fa26

Reply:
changed by splitting up presentation preparation helper in commit 164c2cf56c1b49df88a19008a60f180f5c06fe26